### PR TITLE
Add different groups to Reflectometry (Polref) GUI

### LINKED ISSUE #18119: Add different groups for data reduction
Scientists would like to have different groups of runs, like this:

runs

settingstab

### PR DESCRIPTION
**Pull request by raquelalvarezbanos**
Adds different groups of runs to the Reflectometry GUI.

**To test:**

- Code changes. Note that the first 11 commits come from branch `17817_Label_groups_with_run_title`
- Functional testing: 
  - Open Mantid Plot -> Interfaces -> Reflectometry -> ISIS Reflectometry (Polref).
  - In the "Runs" and "Settings" tabs, check that there are two different groups with the same widgets.
  - Go to "Runs" tab. Check that actions accessible from the menus (i.e. "Reflectometry" and "Edit") apply to the selected table. For instance, loading a .tbl file should populate the table in the current group. The same should happen with the other options (new table, insert row, delete row, etc).
  - Check that both tables are able to reduce data.
  - Enter some values in "Settings". Check that the table in the first groups takes the settings from the first group in "Settings". The same should happen for the second group.

Fixes #18119 

Release notes were updated: https://github.com/mantidproject/mantid/pull/18124/commits/1409ebf8e8d7f9595cd53f5e48fa28c131f0e68f

Documentation will be updated in a separate issue before the next release: https://github.com/mantidproject/mantid/issues/18176

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidQt/CustomInterfaces/src/Reflectometry/ReflMainWindowPresenter.cpp [hunk 1]
```diff
@@ -34,35 +34,41 @@ ReflMainWindowPresenter::ReflMainWindowPresenter(
 ReflMainWindowPresenter::~ReflMainWindowPresenter() {}
 
 /** Returns global options for 'CreateTransmissionWorkspaceAuto'
+*
+* @return :: Index of the group in 'Settings' tab from which to get the options
```
Comment (PranavBahuguna) [inline MantidQt/CustomInterfaces/src/Reflectometry/ReflMainWindowPresenter.cpp]: Shouldn't this be `@param`? I'm surprised that Doxygen didn't pick up on this though. The same goes for all other methods in this class.

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/ReflSettingsPresenter.h [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  std::string getReductionOptions() const override;
+  /// Returns global options for 'Stitch1DMany'
+  std::string getStitchOptions() const override;
+
+private:
+  void createStitchHints();
+  void getExpDefaults();
+  void getInstDefaults();
+  IAlgorithm_sptr createReductionAlg();
+  Instrument_const_sptr createEmptyInstrument(std::string instName);
```
Comment (PranavBahuguna) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/ReflSettingsPresenter.h]: Could this be `Instrument_const_sptr createEmptyInstrument(const std::string &instName);`?

File: MantidQt/CustomInterfaces/src/Reflectometry/ReflMainWindowPresenter.cpp (hunk 1 above)
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/src/Reflectometry/ReflMainWindowPresenter.cpp]: Yes, it should :+1:

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/ReflSettingsPresenter.h (hunk 2 above)
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/ReflSettingsPresenter.h]: :+1:  I've also changed the header to use forward declarations.

**Comment by PranavBahuguna** [comment]
The code seems ok and all functional testing steps produced the expected results.

I've yet again discovered an unrelated issue - when getting default values for quoted strings of values (e.g. `CRho`), the values given are not enclosed in quotes. Thus processing runs fails when attempted with the given defaults. I shall create a separate issue to deal with this.

Anyway once all checks pass, :shipit:

**Comment by raquelalvarezbanos** [comment]
Thanks @PranavBahuguna , it's good that you found that unrelated issue, we still have plenty of time to fix it before the release.

**Comment by raquelalvarezbanos** [comment]
I am not sure what's wrong with the build servers, the code builds on my machine...

**Pull request merged**